### PR TITLE
Enable HMAC on Browser WASM

### PR DESCRIPTION
**Pull request by eerhardt**
This ports the C# implementation of HMAC from https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/mscorlib/system/security/cryptography/hmac.cs, marks the APIs as supported on Browser, and enables HMAC tests on Browser WASM.

Contributes to #40074

The HMAC SubtleCrypto support will come in a separate PR. Currently, we don't have the ability to run both the SubtleCrypto and C# fallback implementations in CI tests. I will work on enabling that before contributing the SubtleCrypto support.

### CODE REVIEW

**Comment by vcsjones** [comment]
Doesn't have to be for this PR, but now that there will be HMAC in browser, a number of other things can probably get re-enabled test-wise, including [HKDF](https://github.com/dotnet/runtime/blob/c0e90753cb5e8c538ab301e1af37f3cf14323bca/src/libraries/System.Security.Cryptography/tests/HKDFTests.cs#L11) and [PBKDF2 (Rfc2898DeriveBytes)](https://github.com/dotnet/runtime/blob/c0e90753cb5e8c538ab301e1af37f3cf14323bca/src/libraries/System.Security.Cryptography/tests/Rfc2898OneShotTests.cs#L11). Those were blocked on HMAC support.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:68]:
```C#
Array.Fill(inner, 0x36);
Array.Fill(outer, 0x5C);
```
?

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:125]:
```suggestion
                Array.Clear(_inner);
                Array.Clear(_outer);
```

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: In the world where we decide that the JIT is allowed to make the same optimizations that the C compiler does, and is allowed to skip writing to memory when it can "prove" that no one will read the value.

I don't expect it to happen, but it's just... one of those things.

Comment (stephentoub) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: > but it's just... one of those things

...that's unnecessary? ;-)

Comment (stephentoub) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: It strikes me as very counterintuitive that we care so much about clearing these arrays that we're going to call a method that avoids optimizations that won't ever be performed, but not enough to clear them if the object is finalized.

Comment (stephentoub) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: > new up an HMACSHA256 with a key, do some stuff, dispose it, memory dump, and then complain if they see any residuals in the memory dump

These arrays aren't pinned.  Even with aggressively calling clear, there's of course no guarantee such a test would succeed.

Comment (eerhardt) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: I don't know how much we want to stray from the original reference source implementation, but AFAICT the `inner` and `outer` arrays are really only used once (inner when the first write happens, outer when finalize happens). Those two arrays could just be `stackalloc`d and initialized when needed. That way they aren't put on the heap.

That doesn't solve the overall problem though, since the key would still need to be stored / cleared.

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: > Those two arrays could just be stackallocd and initialized when needed. That way they aren't put on the heap.

Sounds like a reasonable change to me.

Comment (eerhardt) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs:126]: The drawback to that approach is if someone holds onto a `HMACSHA256` instance and reuses it, we would be re-initializing this data each time the caller used it. Although the initialization isn't very intensive: filling the buffer with a constant and then XOR'ing with the key.

**Review by bartonjs — APPROVED** [review]

**Comment by eerhardt** [comment]
> now that there will be HMAC in browser, a number of other things can probably get re-enabled test-wise, including [HKDF](https://github.com/dotnet/runtime/blob/c0e90753cb5e8c538ab301e1af37f3cf14323bca/src/libraries/System.Security.Cryptography/tests/HKDFTests.cs#L11) and [PBKDF2 (Rfc2898DeriveBytes)](https://github.com/dotnet/runtime/blob/c0e90753cb5e8c538ab301e1af37f3cf14323bca/src/libraries/System.Security.Cryptography/tests/Rfc2898OneShotTests.cs#L11).

`PBKDF2 (Rfc2898DeriveBytes)` is already tracked in https://github.com/dotnet/runtime/issues/40074.

`HKDF` - from what I can tell from the code, everything sitting on top of HMAC is all in C# so it should just be enable tests and mark the APIs as supported on `browser`, right? @bartonjs @GrabYourPitchforks - thoughts on adding this to #40074 as well? Or should I open a new issue for it?

**Comment by GrabYourPitchforks** [comment]
@eerhardt Whatever's easier for you. :) If it were me I'd just add it to the original tracking issue, but if you have a different preferred tracking mechanism I'll defer to your judgment.

Agree that there's no longer any reason to mark APIs as "non-browser" if the underlying primitives are supported.

**Comment by eerhardt** [comment]
OK - I have added `HKDF` to #40074 in the same work item as `PBKDF2`.

**Comment by bartonjs** [comment]
> thoughts on adding this to 40074 as well?

The issue tracks places where we want to plumb through to the "native" implementation, e.g. for PBKDF2 with a sane iteration count (modern numbers should be around 100k) we get to move all of those HMAC resets and context calls to the other side of the message pump, leading to better performance, which leads to being able to use higher iteration counts (which increases the cost of brute forcing).

For HKDF we don't currently have any native implementations, it's written just in terms of platform HMAC (think of HKDF as more a scheme than an algorithm), so it just needs the tests turned on and the UnsupportedOSAttribute removed.

PBKDF2/Rfc2898DeriveBytes would also kinda just work right now, but the static methods would be using the fallback behavior (for platforms that don't expose it as a primitive), and then the algorithm would stay on the list as "do the plumbing for having SubtleCrypto do this faster".  (We have instance methods, too.  No native implementation supports the weird streaming mode our instance methods have, so it's always a fully managed implementation).

**Comment by bartonjs** [comment]
> we really should document what security guarantees we do and do not make when running in these environments

"Your keys are completely insecure. Once you've even thought of having a key somewhere near your browser, it's probably already compromised. Do server-side encryption, use TLS, and leave the clients out of the picture."

Too honest? :smile:

**Pull request merged**
